Replace docs with auto-generated OpenAPI docs

shu8 opened the pull request:
## What does this PR do?
Replaces the manually-written documentation with automatically Swagger-rendered docs from our OpenAPI specification at https://github.com/uclapi/uclapi-openapi/blob/master/uclapi.json.

Unless I've missed anything, the new Swagger docs should have everything that the old docs had **except for** the example snippets for each endpoint.

The old example snippets used to be mostly the same for each endpoint, so if we wanted to keep this, maybe we could just add a single example for each type at the top?

A cURL example is automatically generated by Swagger when you try the API via the UI though.

![image](https://user-images.githubusercontent.com/8850830/104651221-4281cb00-56af-11eb-90b1-1b140a84770c.png)

![image](https://user-images.githubusercontent.com/8850830/104651307-61805d00-56af-11eb-8bef-21e4b0ad0698.png)

Closes #3055. Closes #3054.

## ✅ Pull Request checklist

- [ ] Is this code complete?
- [ ] Are tests done/modified?
- [ ] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
No

## :squirrel: Deploy notes
N/A

## Anything else
Fixes #2964

hueyy commented:
Some of the code examples were a bit special, like the OAuth stuff, which I think it might be useful to keep. Is there any way we can include some of those code snippets, maybe in the description or something?

Feverfew commented:
> Some of the code examples were a bit special, like the OAuth stuff, which I think it might be useful to keep. Is there any way we can include some of those code snippets, maybe in the description of each endpoint or something?

I agree the OAuth workflow should be something we can keep.

A couple of questions I'd like clarified:
1) One thing is that the SDKs should always correspond to what is currently running on the server(s), so the deploy scripts for staging and production should be where the new SDK is pushed from. This sounds like it requires two SDKs, as the code on staging and production is different. Is it possible to take this into account?
2) Related to the above, I'd prefer if the definition was in this repo (and sounds like it will make it easier to take into account the first point). Firstly, this will mean that its easier to enforce that changes in the API are also reflected in changes to the OpenAPI definition at the same time (i.e. in the same PR). Secondly, it sounds like that it will make it easier to meet the requirements of the point I made above.

Otherwise, I like the fact that we've got rid of this many lines of code!

Could we also also have documentation in the `backend/source` directory to explain the recommended workflow in regards to editing the schema?

shu8 commented:
I've updated the spec to add code samples for the OAuth endpoints, and added one example at the top for how to call the roombookings api, which hopefully people will find easy to tweak for the other endpoints: https://github.com/uclapi/uclapi-openapi/commit/c79149fa4148d5c47e61d764f1b59c5db6941cc8

---

I think adding the spec file to this repo makes sense for both the reasons you mentioned, but just a quick (possibly stupid) question: how/when do we currently promote staging to production?

Having separate SDKs should also work, and it might be enough to just create separate branches in the SDK repos and publishing to npm, etc. with different version numbers depending on which one it is.

Feverfew commented:
> I think adding the spec file to this repo makes sense for both the reasons you mentioned, but just a quick (possibly stupid) question: how/when do we currently promote staging to production?

Whenever Zak feels like it, manually. Should be automated with Github Actions, after integration tests on staging are run.

> Having separate SDKs should also work, and it might be enough to just create separate branches in the SDK repos and publishing to npm, etc. with different version numbers depending on which one it is.
Yeah the workflow I'm thinking of is.
1. Master is updated.
2. Deployment to staging. Once deployed, job is triggered to create new staging SDK(s) based on OpenAPI spec in the repo and push them to NPM automatically (potentially with some smoke tests?).
3. Same for production, but obviously a different SDK. Whether it's a different branch or repo, I'm not too worried about.

Feverfew commented:
As discussed in the meeting once the OpenAPI spec is added to this repo and the documentation is built from it, then I'm happy to review and merge this. We can deal with SDK stuff in a separate PR

Feverfew commented on frontend/src/pages/Documentation.jsx:
```diff
@@ -1,58 +1,17 @@
-  </StylesProvider>
+  <div className='vertical-padding'>
+    <NavBar isScroll={false} />
+    <SwaggerUI url="https://cdn.jsdelivr.net/gh/uclapi/uclapi-openapi/uclapi.json" />
```
Can this be built statically? The reason I ask is, firstly for performance reasons, but more importantly is for consistency. At what point is `https://cdn.jsdelivr.net/gh/uclapi/uclapi-openapi/uclapi.json` updated? It seems like either the documentation should be generated as part of a build step or be part of the statics?

shu8 replied:
Yeah I agree statically building is best -- I should have really marked this as WIP after our meeting sorry! 

I haven't had much time to look into the best way to build it statically and join it with the webpack build yet, but I'll try looking into it ASAP - openapi-generator has a [dynamic-html docs generation option](https://openapi-generator.tech/docs/generators/dynamic-html) which looks promising for us!

shu8 replied:
It's part of the build process now!

Feverfew reviewed (changes requested):
One question that I'd like to hash out before approval

shu8 commented:
@Feverfew I've:

- added some sphinx docs for the openapi documentation process,
- moved the openapi docs rendering to be a build-time step so it is always in sync with the current deployment, and
- updated the spec to add descriptions to all the fields (which we had in the existing docs but I forgot to copy over when I did the openapi spec originally!)

if it looks good then I will move onto the SDK deployment work and archive the old `uclapi-openapi` repo (https://github.com/uclapi/uclapi-openapi)!

Feverfew reviewed (dismissed):
LGTM let's get this in staging.

greenfrogs commented:
We need to wait for staging to be fixed first, and #3052 to merge. Possibly also adding the new docs from #3052 (otherwise we are going to have some really annoying merge conflicts on the other request).

shu8 commented:
> Possibly also adding the new docs from #3052 (otherwise we are going to have some really annoying merge conflicts on the other request).

I'll add those into this PR now, so once workspaces is merged we can get this in quicker!

shu8 commented:
I've added the docs from #3052 into this PR but the review got dismissed with the new push! Please could either of you have another quick look @greenfrogs @Feverfew!

shu8 commented:
sorry for the delay but all merge conflicts and tests should be fixed now for the final re-review! @greenfrogs @Feverfew

zipy124 commented:
@uclapi/ucl-api-team are we ready to get this into master now?

shu8 commented:
@zipy124 I think it is -- the last review got dismissed because of all the merge conflicts I fixed and the new workspaces docs but nothing else has changed! 

I'll resolve this package-lock.json conflict now too!

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+      "email" : "isd.apiteam@ucl.ac.uk",
+      "url" : "https://uclapi.com"
+    },
+    "version" : "1.0.0-oas3",
```
nit: do we want to match this version to some sort of release of the API?

shu8 replied:
Yes! We can do that along with #3379!

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "email" : {
+            "type" : "string",
+            "example" : "ucachun@ucl.ac.yk",
```
nit: ac.uk

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "capacity" : {
+            "type" : "integer",
+            "format" : "int64",
```
ooc: any reason for `int64` instead of `number`?

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+                  },
+                  "lng" : {
+                    "type" : "number",
+                    "example" : "-0.13366",
```
nit: can these numbers be outside of quotes like above (or do floats cause problems)?

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "siteid" : {
+            "type" : "number",
+            "example" : 86.0,
```
super nit: remove .0

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "roomid" : {
+            "type" : "number",
+            "example" : "433",
```
nit: number in quotes

shu8 replied:
this is also actually a string, updated!

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+                  },
+                  "lng" : {
+                    "type" : "string",
+                    "example" : "-0.127864",
```
nit: number in quotes

shu8 replied:
Turns out we actually return lat/long as a string! that seems intentional, maybe for storage or precisions concerns?

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "last_trigger_timestamp" : {
+            "type" : "string",
+            "example" : "2018-02-15T22:42:28+00:00",
```
ooc: do we return time zone here? Our API seems to be really inconsistent 😅

shu8 replied:
Yeah it look like we return exactly what comes from occupeye

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          },
+          "x_pos" : {
+            "type" : "string",
+            "example" : "32432.0",
```
This might be a mistake in the API but does this return a string rather than a number? (same for `survey_device_id`)

shu8 replied:
Yeah looks like we're returning a string in the API, again using the value straight from Occupeye

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+      "historical_survey_data" : {
+        "properties" : {
+          "sensor_id" : {
+            "type" : "integer",
```
nit: integer instead of number (same with `averages`)

shu8 replied:
I think it might be better to be more specific and stick with `integer` here?

shu8 replied:
Looking through the code we have a lot of fields with types that I don't really get, e.g., `weeknumber` is a `FloatField` for `/roombookings`, and `BigIntegerField` for `/timetable`.

we could probably update a lot of our types -- what do you think? if we did #3060 this could go along with it.

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+      "historical_survey" : {
+        "properties" : {
+          "survey_id" : {
+            "type" : "integer",
```
nit: int

shu8 replied:
same with this! what do you think?

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+                "properties" : {
+                  "width" : {
+                    "type" : "string",
+                    "example" : "28329.0",
```
nit: same as above string for number value

shu8 replied:
I think this is a string because it's a HTML SVG so all the attributes would be strings

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+          "required" : true,
+          "schema" : {
+            "type" : "string",
+            "example" : 1
```
nit: is this a string? is this a good example?

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+                  "Invalid date/time" : {
+                    "value" : {
+                      "ok" : false,
+                      "error" : "date/time isn't formatted as suggested in the docs"
```
nit: I think this is a mistake in the API, I believe we format date/time as datetime

shu8 replied:
updated error message to be more clear: `start_datetime/end_datetime isn't formatted as suggested in the docs`

greenfrogs commented on uclapi.openapi.json:
```diff
@@ -0,0 +1,4256 @@
+  },
+  "servers" : [ {
+    "url" : "https://uclapi.com",
+    "description" : "Deployment server"
```
nit: can we call this `Production server`?

greenfrogs reviewed (approved):
@shu8 sorry for the very late review. This is an excellent piece of work and will make it so much easier to support and use the API in the future.

I have some really small comments but they are mostly only nits so I am happy to approve for merging when you feel it is best. Lots of my comments are likely problems with the underlying API or weird inconsistencies which have risen as a result of difficulty in seeing datatypes/phrasing for the entire API. Feel free to not address these and I can raise issues to address them in the future.

shu8 commented:
Thanks @greenfrogs! Yeah those all seem like things we should fix!

I should have time to go over these (and go through all the endpoints fully once more to see if there is any other weirdness!) over the next week but it might be good to get this into master/staging and I can submit another PR for all the fixes?

zipy124 commented:
> Thanks @greenfrogs! Yeah those all seem like things we should fix!
> 
> I should have time to go over these (and go through all the endpoints fully once more to see if there is any other weirdness!) over the next week but it might be good to get this into master/staging and I can submit another PR for all the fixes?

I'm good with this decision!
